service/src/main/java/skills/metrics/builders/project/UserAchievementsMetricsBuilder.groovy

```diff
@@ -0,0 +1,145 @@
+
+        int totalNumItems = userAchievedRepo.countForAchievementNavigator(
+                projectId, usernameFilter, from, to, skillNameFilter, minLevel, achievementTypesWithoutOverall, allNonOverallTypes, includeOverallType)
+        if (totalNumItems == 0) {
+            return new UserAchievementsRes(totalNumItems: 0, items: [])
+        }
+        List<Object[]> achievements = userAchievedRepo.findAllForAchievementNavigator(
+                projectId, usernameFilter, from, to, skillNameFilter, minLevel, achievementTypesWithoutOverall, allNonOverallTypes, includeOverallType, pageRequest)
+
+        List items = achievements.collect {
+            UserAchievement userAchievement = it[0]
+            SkillDef skillDef = it[1]
+            return buildMetricUserAchievement(userAchievement, skillDef)
+        }
+        return new UserAchievementsRes(totalNumItems: totalNumItems, items: items)
+    }
+
+    private PageRequest getPageRequest(String projectId, String chartId, Map<String, String> props) {
+
+
+        MetricsPagingParamsHelper metricsPagingParamsHelper = new MetricsPagingParamsHelper(projectId, chartId, props)
+        int currentPage = metricsPagingParamsHelper.currentPage
+        int pageSize = metricsPagingParamsHelper.pageSize
+        Boolean sortDesc = metricsPagingParamsHelper.sortDesc
+        String sortBy = metricsPagingParamsHelper.sortBy
+
+        if (!supportedSortBy.contains(sortBy)) {
+            throw new SkillException("Metrics[${chartId}]: Invalid value [${props[MetricsPagingParamsHelper.PROP_SORT_BY]}] for [${MetricsPagingParamsHelper.PROP_SORT_BY}] property. Suppored values are ${supportedSortBy}", projectId)
+        }
+
+        if (sortBy == "userName") {
+            // looks like there is a bug in the Spring Data's
+            //  org.springframework.data.jpa.repository.query.QueryUtils.java
+            // where aliases are not respected in some JPQL joins and the default alias (first table in the join) is added
+            // no matter what is provided.
+            // It maybe fixed by https://jira.spring.io/browse/DATAJPA-1406 even though they scope it to Native Queries
+            // the issue is present on JPQL queries as well
+            //
+            // This is a workaround to "disable" aliasing
+            Sort sort = JpaSort.unsafe(sortDesc ?  DESC : ASC, "(uAttrs.userIdForDisplay)")
+            return PageRequest.of(currentPage, pageSize, sort)
+        }
+        return PageRequest.of(currentPage, pageSize, sortDesc ?  DESC : ASC, sortBy)
+    }
+    private MetricUserAchievement buildMetricUserAchievement(UserAchievement achievement, SkillDef skillDef) {
+
+        MetricUserAchievement res = new MetricUserAchievement(
+                achievedOn: achievement.achievedOn.time,
+                userId: achievement.userId,
+                userName: achievement.userId)
+
+        if (!achievement.skillId) {
```

Comment:
Could be simplified by moving the !skillId values into the default construction on line 127 and changing the if/else to only a if(skillId) block.